integrations/access/common/access_monitoring_rules.go

```diff
@@ -0,0 +1,193 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package common
+
+import (
+	"context"
+	"maps"
+	"sync"
+
+	"github.com/gravitational/trace"
+
+	accessmonitoringrulesv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessmonitoringrules/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/integrations/access/common/teleport"
+	"github.com/gravitational/teleport/integrations/lib/logger"
+)
+
+const (
+	// defaultAccessMonitoringRulePageSize is the default number of rules to retrieve per request
+	defaultAccessMonitoringRulePageSize = 10
+)
+
+// AccessMonitoringRuleHandler stores a cache of Access Monitoring Rules for use with Access Request routing in plugins.
+// Must be initialised by calling InitAccessMonitoringRulesCache, a watcher on Acccess Monitoring Rules must pass in new rules using HandleAccessMonitoringRule.
+type AccessMonitoringRuleHandler struct {
+	accessMonitoringRules AmrMap
+
+	apiClient  teleport.Client
+	pluginType string
+
+	ruleAppliesCallback        func(amr *accessmonitoringrulesv1.AccessMonitoringRule) bool
+	fetchRecipientCallback     func(ctx context.Context, recipient string) (*Recipient, error)
+	matchAccessRequestCallback func(expr string, req types.AccessRequest) (bool, error)
+}
+
+// AmrMap is a concurrent map for access monitoring rules.
+type AmrMap struct {
```

Comment:
```suggestion
type AMRMap struct {
```

Initialisms should be all caps.

That said, as I've mentioned before "AMR" is not a commonly understood initialism. I would be more explicit.

WDYT about moving this code out of `common` and into an `accessmonitoring` package? `accessmonitoring.RuleMap` is a lot more descriptive of a name than `common.AmrMap`.

Reply:
That sounds way better